packages/mui-material/src/ButtonGroup/ButtonGroup.js

```diff
@@ -81,106 +88,106 @@ const ButtonGroupRoot = styled('div', {
   }),
   [`& .${buttonGroupClasses.grouped}`]: {
     minWidth: 40,
-    '&:not(:first-of-type)': {
-      ...(ownerState.orientation === 'horizontal' && {
-        borderTopLeftRadius: 0,
-        borderBottomLeftRadius: 0,
-      }),
-      ...(ownerState.orientation === 'vertical' && {
-        borderTopRightRadius: 0,
-        borderTopLeftRadius: 0,
+    '&:hover': {
+      ...(ownerState.variant === 'contained' && {
+        boxShadow: 'none',
       }),
-      ...(ownerState.variant === 'outlined' &&
-        ownerState.orientation === 'horizontal' && {
-          marginLeft: -1,
-        }),
-      ...(ownerState.variant === 'outlined' &&
-        ownerState.orientation === 'vertical' && {
-          marginTop: -1,
-        }),
     },
-    '&:not(:last-of-type)': {
-      ...(ownerState.orientation === 'horizontal' && {
-        borderTopRightRadius: 0,
-        borderBottomRightRadius: 0,
+    ...(ownerState.variant === 'contained' && {
+      boxShadow: 'none',
+    }),
+  },
+  [`& .${buttonGroupClasses.firstButton},.${buttonGroupClasses.middleButton}`]: {
```

Comment:
@ZeeshanTamboli Instead of exposing the className, how about using just `data-*` attribute like:
- data-first-child
- data-last-child

No need for middle because we can use CSS `:not` to target middle children like this:
`.${buttonGroupClasses.root}:not([data-first-child], [data-last-child])`

Reply:
@siriwatknp I thought about this and also referred Joy UI's ButtonGroup component code. The reason I added middle class is because then the `:not` selector will select the wrapping elements. As an example, in the case of the Tooltip shown in the CodeSandbox provided in the description it will select the Tooltip's `span` elements for:
```jsx
<ButtonGroup>
  <Tooltip title="tooltip">
    <Button>Enabled</Button>
  </Tooltip>
  <Tooltip title="tooltip">
    <span>
      <Button disabled>Disabled</Button>
    </span>
  </Tooltip>
  <Tooltip title="tooltip">
    <span>
      <Button disabled>Disabled</Button>
    </span>
  </Tooltip>
</ButtonGroup>
```
And `span` is also needed to show the tooltip on disabled buttons - https://mui.com/material-ui/react-tooltip/#disabled-elements. Related issue #29514. The `:not` selector, however, won't target nested elements, such as a button inside a tooltip.

Regarding exposing classes, offering customization options to users might be a good idea. However, if desired, I could also use `data-first-child`, `data-middle-child`, and `data-last-child` attributes. Let me know your preference.

Reply:
Alright, let's try `data-middle-child` instead of classes so that we don't expose too much API at this point.

Reply:
@siriwatknp Done.

Reply:
@siriwatknp, why not classes? We tend to use classes everywhere in Material UI for styling.

Reply:
I re-added classes now.